### ISSUE TITLE
[css-position-3] Sticky position and margins (should margins be considered at all?).

### ISSUE THREAD

**Issue by bfgeek** [w3c/csswg-drafts#9052]
https://drafts.csswg.org/css-position-3/#stickypos-insets

> For each side of the box, if the corresponding [inset property](https://drafts.csswg.org/css-position-3/#inset-properties) is not [auto](https://drafts.csswg.org/css-position-3/#valdef-top-auto), and the corresponding [border edge](https://drafts.csswg.org/css-box-4/#border-edge) of the box would be outside the corresponding edge of the [sticky view rectangle](https://drafts.csswg.org/css-position-3/#sticky-view-rectangle), then the box must be visually shifted ([as for relative positioning](https://drafts.csswg.org/css-position-3/#relpos-insets)) to be inward of that sticky view rectangle edge, insofar as it can while its [position box](https://drafts.csswg.org/css-position-3/#position-box) remains contained within its [containing block](https://drafts.csswg.org/css-display-4/#containing-block). The position box is its [margin box](https://drafts.csswg.org/css-box-4/#margin-box), except that for any side for which the distance between its [margin edge](https://drafts.csswg.org/css-box-4/#margin-edge) and the corresponding edge of its containing block is less than its corresponding [margin](https://drafts.csswg.org/css-box-4/#margin), that distance is used in place of that margin.

The spec and implementations are currently a little scattered when it comes to how margins should be treated for sticky-pos elements. It's not clear if all the interactions have been thought through completely.

1. Implementations (and the spec I think?) currently use border-box rect to determine if something is overconstrained (e.g. `position: sticky; left: 100px; right: 100px;`.  Is this intentional/correct?

2. Implementations currently differ on if auto-margins should be considered. For example: https://www.software.hixie.ch/utilities/js/live-dom-viewer/?saved=11848
```
<!DOCTYPE html>
<div style="overflow: scroll; width: 100px; height: 100px;">
  <div style="margin: 20px; border: solid; width: 100px; height: 100px;">
    <div style="position: sticky; left: 0; width: 10px; height: 10px; margin-right: auto; background: lime;"></div>
  </div>
</div>
```

Per-spec Firefox is currently correct - however this might be surprising to web-developers as it'll effectively opt-out the element being sticky by default if they are just using auto-margins to center content.

3. Unclear from the spec how collapsing margins should be considered in the calculations above. Currently no implementations appear to look at collapsing margins.

4. The sticky view rectangle is currently defined as the border-box which seems to match what web-developers would expect. E.g. you don't expect the sticky-element to stick based on some unseen margin. Do the margins in the rest of the calculation add create more confusion than value? E.g. margins are often used for aligning content, if you use margins in this way and then make the element sticky-pos then it'll likely have un-expected behaviour (see auto-margins above as one example).

TL;DR Should the "position box" just be the border-box instead? (rather than the margin-box).

**Comment by flackr** [comment]
> 1. Implementations (and the spec I think?) currently use border-box rect to determine if something is overconstrained (e.g. `position: sticky; left: 100px; right: 100px;`.  Is this intentional/correct?

I think this is correct, since margins are not maintained from the sticky view rectangle. For example, consider the following ([demo](https://jsbin.com/gejehiy/edit?html,output)):
```html
<style>
.scroller {
  height: 400px;
  overflow: auto;
  border: 1px solid black;
}
.sticky {
  margin: 250px 0;
  height: 200px;
  position: sticky;
  background: green;
  top: 0;
  bottom:0;
}
.spacer {
  height: 200px;
  background: #ccf;
}
</style>
<div class="scroller">
  <div class="spacer"></div>
  <div class="sticky"></div>
  <div class="spacer"></div>
</div>
```

The sticky element is not overconstrained. It can in fact be pushed up / down into the margin reserved between it and the adjacent spacers until it is 250px away from the top/bottom of the container / scroller. That said I still find it confusing that we try to preserve its margins from the container.

> 3. Unclear from the spec how collapsing margins should be considered in the calculations above. Currently no implementations appear to look at collapsing margins.

IMO if we consider margins we should allow for them to collapse.

> 4. The sticky view rectangle is currently defined as the border-box which seems to match what web-developers would expect. E.g. you don't expect the sticky-element to stick based on some unseen margin. Do the margins in the rest of the calculation add create more confusion than value? E.g. margins are often used for aligning content, if you use margins in this way and then make the element sticky-pos then it'll likely have un-expected behaviour (see auto-margins above as one example).
> 
> TL;DR Should the "position box" just be the border-box instead? (rather than the margin-box).

I agree, I think it would be simpler if the sticky adjustment were only constrained by the border box, and this might enable margins to be used to set the initial position. However, given that we seem to have compat on this part (with non-auto margins) there's a chance this will be breaking. In particular when the sticky element is first I could see developers being surprised if it gets pushed up into its top margin, or if last being pushed down into its bottom margin.

**Comment by dholbert** [comment]
FWIW, I filed these Gecko bugs:
https://bugzilla.mozilla.org/show_bug.cgi?id=1844435 on ignoring margins for sticky-positioning (pending evidence that that's web-compatible)
https://bugzilla.mozilla.org/show_bug.cgi?id=1844438 as a smaller intermediate step of ignoring auto margins (which is a smaller step that seems more clearly web-compatible given that that's what Blink/WebKit already do)

**Comment by johannesodland** [comment]
The dual implication of margin on sticky elements has [been an issue](https://github.com/w3c/csswg-drafts/issues/2496) and if it's web compatible it will be good to clean up that ambiguity. However, the margin has been our only way of controlling the "sticky constraint rectangle" (or the "sticky view rectangle" as it was called in the discussion). 

[jonjohnjohnson](https://github.com/jonjohnjohnson) has described a couple of use cases for controlling the SCR in [the issue](https://github.com/w3c/csswg-drafts/issues/2496), but I'll add another here for reference:

<details>
  <summary>Use case for controlling the sticky constraint rectangle</summary>
  
  ### Article with sticky headers
  An article has multiple sections, and each section has a sticky header element.
  When the user reach the bottom of each section, the header element should be pushed out of view, 
  before the next section and header element is scrolled into view. 
  
 Currently, the solution for limiting the sticky constraint rectangle is to set a margin-bottom on the sticky element, and then negating that margin on the following sibling:

  ```html
  <article>
    <section>
      <header>...</header>
      ...
    </section>
    <section>
      <header>...</header>
      ...
    </section>
  </article>
  ```
  ```css
  header {
    position: sticky;
    margin-bottom: 90vh; /* Limit SCR */
  }
  
  header + * {
    margin-top: -90vh; /* Negate effect of margin bottom */ 
  }
  ```
</details>

As it's been resolved to ignore margins in sticky-pos calculations, can we find another way to control the SCR? Something similar to `scroll-margin`?

I know we have content that use margin to control the SCR, and we will have to find other solutions when this is implemented. The only solution (that I can think of) that is supported in browsers today is to add padding to the block level ancestor, and then negate that in the child nearest to the padding edge. That feels even worse than what we do today.

**Comment by jonjohnjohnson** [comment]
I respectfully disagree with this resolution and hope it ends up not web-compatible, primarily on the grounds that negative margins have allowed sticky elements to stay "stuck" (with fixed-like behavior) in webkit/gecko during overscroll animations.

See https://github.com/w3c/csswg-drafts/issues/8309#issuecomment-1443059377

I also second much of @johannesodland in https://github.com/w3c/csswg-drafts/issues/9052#issuecomment-1809658376

**Comment by johannesodland** [comment]
Returning to alternative solutions for defining the "sticky constraint rectangle":

Currently the sticky positioned element is shifted to stay within the sticky view rectangle insofar its possible while staying within its containing block inset by its margins (simplified). In other words, sticky pos operates with two insets, one inset from its scrollport, and one inset from its containing block. If margins are no longer to be considered, authors have no way of controlling that second inset.

I'm wondering if the box insets could be extended to support multiple named insets, to give a way of controlling the inset from the containing block?

```css
/* 
Make element sticky against scrollport 
insofar as it can while staying 20px within its containing block 
*/
position: sticky;
inset-block: 
  scrollport 0px,
  containing-block 20px;
```

This could solve issue one in https://github.com/w3c/csswg-drafts/issues/2496.

(A `position-container` property as mentioned in https://github.com/w3c/csswg-drafts/issues/9868 could solve issue 2 by allowing you to change the containing block.)
